storage/redis/peer_store.go

```diff
@@ -0,0 +1,787 @@
+			if numWant == 0 {
+				break
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+			numWant--
+		}
+	} else {
+		// Append as many seeders as possible.
+		for _, pk := range con_seeders {
+			if numWant == 0 {
+				break
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+			numWant--
+		}
+
+		// Append leechers until we reach numWant.
+		if numWant > 0 {
+			announcerPK := newPeerKey(announcer)
+			for _, pk := range con_leechers {
+				if pk == announcerPK {
+					continue
+				}
+
+				if numWant == 0 {
+					break
+				}
+
+				peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+				numWant--
+			}
+		}
+	}
+
+	APResult := ""
+	for _, pr := range peers {
+		APResult = fmt.Sprintf("%s Peer:[ID: %s, IP: %s(AddressFamily: %s), Port %d]", APResult, pr.ID.String(), pr.IP.String(), IPver, pr.Port)
+	}
+	log.Debug("storage: AnnouncePeers result", log.Fields{
+		"peers": APResult,
+	})
+
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	resp.InfoHash = ih
+	IPver := addressFamily.String()
+	encoded_ih := ih.String()
+	encoded_ih_L := IPver + "_L_" + encoded_ih // key
+	encoded_ih_S := IPver + "_S_" + encoded_ih // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechers_len, err := conn.Do("HLEN", encoded_ih_L)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encoded_ih_L,
+			"error": err,
+		})
+		return
+	}
+	l_len := leechers_len.(int64)
+
+	seeders_len, err := conn.Do("HLEN", encoded_ih_S)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encoded_ih_S,
+			"error": err,
+		})
+		return
+	}
+	s_len := seeders_len.(int64)
+
+	if l_len == 0 && s_len == 0 {
+		return
+	}
+
+	resp.Incomplete = uint32(l_len)
+	resp.Complete = uint32(s_len)
+
+	return
+}
+
+// collectGarbage deletes all Peers from the PeerStore which are older than the
+// cutoff time.
+//
+// This function must be able to execute while other methods on this interface
+// are being executed in parallel.
+func (ps *peerStore) collectGarbage(cutoff time.Time) error {
+	select {
+	case <-ps.closed:
+		return nil
+	default:
+	}
+
+	shards := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	cutoffUnix := cutoff.UnixNano()
+	start := time.Now()
+
+	for _, shard := range shards {
+		infohashes_list, err := conn.Do("HKEYS", shard) // key
+		if err != nil {
+			return err
+		}
+		infohashes := infohashes_list.([]interface{})
+
+		for _, ih := range infohashes {
+			ih_str := string(ih.([]byte))
+
+			ih_list, err := conn.Do("HGETALL", ih_str) // field
+			if err != nil {
+				return err
+			}
+			con_ih := ih_list.([]interface{})
+
+			if len(con_ih) == 0 {
```

Comment:
This looks racy - what if another connection added a peer to the infohash between our `HGETALL` and this check?

Reply:
I do think it's ok under this circumstances as `lost one or serveral peers` during this interval doesn't affect the overall Peers conditions(there are still many other peers we can acquire).